src/bio/bio_wal.c

```diff
@@ -979,27 +979,628 @@ flush_wal_header(struct bio_meta_context *mc)
 	return write_header(mc, mc->mc_wal, hdr, sizeof(*hdr), &hdr->wh_csum);
 }
 
-int
-bio_wal_replay(struct bio_meta_context *mc,
-	       int (*replay_cb)(struct umem_action *actv, unsigned int act_nr),
-	       unsigned int max_replay_nr)
+static int
+load_wal(struct bio_meta_context *mc, char *buf, unsigned int max_blks, uint64_t tx_id)
 {
-	/* TODO */
+	struct wal_super_info	*si = &mc->mc_wal_info;
+	unsigned int		 tot_blks = si->si_header.wh_tot_blks;
+	unsigned int		 blk_bytes = si->si_header.wh_blk_bytes;
+	struct bio_sglist	 bsgl;
+	struct bio_iov		*biov;
+	d_sg_list_t		 sgl;
+	d_iov_t			 iov;
+	unsigned int		 nr_blks, blks, off;
+	bio_addr_t		 addr = { 0 };
+	int			 iov_nr, rc;
+
+	d_iov_set(&iov, buf, max_blks * blk_bytes);
+	sgl.sg_iovs = &iov;
+	sgl.sg_nr = 1;
+	sgl.sg_nr_out = 0;
+
+	/* Read in 1MB sized IOVs */
+	nr_blks = (1UL << 20) / blk_bytes;
+	D_ASSERT(nr_blks > 0);
+	iov_nr = (max_blks + nr_blks - 1) / nr_blks + 1;
+	rc = bio_sgl_init(&bsgl, iov_nr);
+	if (rc)
+		return rc;
+
+	off = id2off(tx_id);
+	while (max_blks > 0) {
+		biov = &bsgl.bs_iovs[bsgl.bs_nr_out];
+
+		bio_addr_set(&addr, DAOS_MEDIA_NVME, off2lba(si, off));
+		blks = min(max_blks, nr_blks);
+		if (off + blks > tot_blks)
+			blks = tot_blks - off;
+		bio_iov_set(biov, addr, blks * blk_bytes);
+
+		bsgl.bs_nr_out++;
+		max_blks -= blks;
+		off += blks;
+		if (off == tot_blks)
+			off = 0;
+		D_ASSERT(bsgl.bs_nr_out <= iov_nr);
+	}
+	/* Adjust the bs_nr for following bio_readv() */
+	bsgl.bs_nr = bsgl.bs_nr_out;
+
+	rc = bio_readv(mc->mc_wal, &bsgl, &sgl);
+	bio_sgl_fini(&bsgl);
+
+	return rc;
+}
+
+static int
+verify_tx_hdr(struct wal_super_info *si, struct wal_trans_head *hdr, uint64_t tx_id)
+{
+	bool	committed = false;
+
+	if (wal_id_cmp(si, tx_id, si->si_commit_id) <= 0)
+		committed = true;
+
+	if (hdr->th_magic != WAL_HDR_MAGIC) {
+		D_CDEBUG(committed, DLOG_ERR, DB_IO, "Mismatched WAL head magic, %x != %x\n",
+			 hdr->th_magic, WAL_HDR_MAGIC);
+		return committed ? -DER_INVAL : 1;
```

Comment:
what the magic ret '1' means here?

Reply:
1 means replay is done since we hit an uncommitted entry in the WAL.